modules/standard/IO.chpl

```diff
@@ -8095,39 +8097,93 @@ private proc readBytesOrString(ch: fileReader, ref out_var: ?t, len: int(64)) :
 
     try ch.lock(); defer { ch.unlock(); }
 
-    var binary:uint(8) = qio_channel_binary(ch._channel_internal);
-    var byteorder:uint(8) = qio_channel_byteorder(ch._channel_internal);
+    // note the current channel position
+    var pos = qio_channel_offset_unlocked(ch._channel_internal);
+
+    // adjust uselen according to the channel's region,
+    // which cannot change during this read
+    var end = qio_channel_end_offset_unlocked(ch._channel_internal);
+    if end != max(int(64)) {
+      // if the channel had an end position set, compute distance to it
+      var channelLen = end - pos;
+      if channelLen < uselen {
+        uselen = channelLen + 1; // +1 to get an EOF back even if already there
+      }
+    }
 
-    if binary {
-      err = qio_channel_read_string(false, byteorder,
-                                    iostringstyleInternal.data_toeof:int(64),
-                                    ch._channel_internal, tx,
-                                    lenread, uselen);
-    } else {
-      var save_style: iostyleInternal = ch._styleInternal();
-      var style: iostyleInternal = ch._styleInternal();
-      style.string_format = QIO_STRING_FORMAT_TOEOF;
-      ch._set_styleInternal(style);
-
-      if t == string {
-        err = qio_channel_scan_string(false,
-                                      ch._channel_internal, tx,
-                                      lenread, uselen);
+    // compute a guess as to the size to read based on the file length.
+    // this is only a guess & it's possible it will be out of date
+    // by the time we actually read.
+    var guessReadSize = 0;
```

Comment:
I would like `guessReadSize` to be calculated in a separate helper function, like `uselen`.